pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,309 @@
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	metricsScraper "kubevirt.io/kubevirt/pkg/monitoring/domainstats/downwardmetrics"
+)
+
+const (
+	maxConnectAttempts   = 6
+	maxRequestsPerSecond = 5
+	maxRequestsBurst     = 1 // must be >= 1, otherwise `rateLimiter.Wait()` will fail
+	invalidRequest       = "INVALID REQUEST\n\n"
+	emptyMetrics         = "<metrics><!-- host metrics not available --><!-- VM metrics not available --></metrics>"
+)
+
+// This is a compile-time assertion to ensure that `maxRequestsBurst` is >= 1, otherwise `rateLimiter.Wait()` will fail
+// (will also fail for `maxRequestsBurst` > 256)
+const _ = uint8(maxRequestsBurst - 1)
+
+func RunDownwardMetricsVirtioServer(ctx context.Context, nodeName, channelSocketPath, launcherSocketPath string) error {
+	report, err := newMetricsReporter(nodeName, launcherSocketPath)
+	if err != nil {
+		return err
+	}
+
+	server := downwardMetricsServer{
+		rateLimiter:        rate.NewLimiter(maxRequestsPerSecond, maxRequestsBurst),
+		maxConnectAttempts: maxConnectAttempts,
+		virtioSerialSocket: channelSocketPath,
+		reportFn:           report,
+	}
+	go server.start(ctx)
+	return nil
+}
+
+type metricsReporter func() (*api.Metrics, error)
+
+func newMetricsReporter(nodeName, launcherSocketPath string) (metricsReporter, error) {
+	exists, err := diskutils.FileExists(launcherSocketPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.New("virt-launcher socket not found")
+	}
+
+	scraper := metricsScraper.NewReporter(nodeName)
+
+	return func() (*api.Metrics, error) {
+		return scraper.Report(launcherSocketPath)
+	}, nil
+}
+
+// The DownwardMetrics server is special, in the sense that the socket is created
+// by QEMU as a server (listen) and the DownwardMetrics server connects to it.
+// Once the connection is established, the application inside the VM has to open
+// the character device and send the requests, and each request is handled sequentially.
+// Only one application can open the character device at the same time, and reading/writing
+// the device blocks the application
+type downwardMetricsServer struct {
+	rateLimiter        *rate.Limiter
+	maxConnectAttempts uint
+	virtioSerialSocket string
+	reportFn           metricsReporter
+}
+
+func (s *downwardMetricsServer) start(ctx context.Context) {
+	conn, err := connect(ctx, s.virtioSerialSocket, s.maxConnectAttempts)
+	if err != nil {
+		log.Log.Reason(err).Error("failed to connect to virtio-serial socket")
+		return
+	}
+
+	s.serve(ctx, conn)
+}
+
+func (s *downwardMetricsServer) serve(ctx context.Context, conn net.Conn) {
+	defer func(conn net.Conn) {
+		err := conn.Close()
+		if err != nil {
+			log.Log.Reason(err).Warning("virtio-serial failed to close the connection")
+		}
+	}(conn)
+
+	type reqResult struct {
+		request string
+		err     error
+	}
+
+	// We must provide a space in the buffer to prevent the wait-for-request goroutine from blocking
```

Comment:
Sorry I can't follow. How does a buffer help here? If the goroutine blocks it should unblock once the select tries to read from the chan? Or should this allow to unblock the routine when the select below returns if `ctx->Done()`? If so can you add a test case for it?

Reply:
ACK, thanks for clarifying.